File: my-folder/0141-linked-list-cycle/solution.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function hasCycle(head: ListNode | null): boolean {
    return traverse(head, head?.next);
};

function traverse(s: ListNode | null, f: ListNode | null): boolean {
    if (s === null || f === null) return false;

    if (s === f) return true;

    return false || traverse(s.next, f?.next?.next);
}

File: my-folder/0141-linked-list-cycle/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        t = h = head

        while t and h and h.next:
            h = h.next.next
            t = t.next

            if t == h:
                return True
        
        return False
